File: ex23.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        r = []

        for i in range(0, len(lists)):
            a = lists[i]

            while a != None:
                r.append(a.val)
                a = a.next

        r.sort()
            
        return r

File: code/ex23.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<pair<int, ListNode*>, vector<pair<int, ListNode*>>, greater<pair<int, ListNode*>>> pq;
        for (ListNode* node : lists) {
            if (node != NULL) {
                pq.push({node->val, node});
            }
        }
        
        ListNode* head = new ListNode();
        ListNode* p = head;
        while (!pq.empty()) {
            int val = pq.top().first;
            ListNode* node = pq.top().second;
            pq.pop();
            if (node->next != NULL)    pq.push({node->next->val, node->next});
            p->next = node;
            p = p->next;
            p->next = NULL;
        }
        
        return head->next;
    }
};